### PR TITLE
nested-datatypes Blog Post

### PR DESCRIPTION
**Pull request by hgoldstein95**
I think we're just about good to go with this post. I had to do a little bit of manual clean-up after pandoc, so I'm not sure that the formatting is 100% perfect.

If possible I'd like to post this this afternoon, so if someone can give this a quick read before then that would be great.

### CODE REVIEW

File: blog/2020-12-04-nested-datatypes/assets/Nested.lhs [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+However, instead of using a nested datatype, it is also possible to use a
+type-level natural number to bound the scopes of bound variables, as shown
+in [this implementation](https://github.com/sweirich/lennart-lambda/blob/master/lib/DeBruijnScoped.lhs).
+
+* [Finger trees](http://www.staff.city.ac.uk/~ross/papers/FingerTree.html)
+
+Haskell's implementation of the
+[sequence](https://hackage.haskell.org/package/containers-0.6.4.1/docs/Data-Sequence.html)
+data structure is built on FingerTrees. In the module
+[DFinger.lhs](DFinger.lhs) I've sketched out a nat-indexed replacement to the
```
Comment (sweirich) [inline blog/2020-12-04-nested-datatypes/assets/Nested.lhs:623]: This link is unbound. Should be to https://github.com/sweirich/dth/blob/master/nested-datatypes/DFinger.lhs

**Review by sweirich — CHANGES_REQUESTED** [review]

File: blog/2020-12-04-nested-datatypes/2020-12-04-nested-datatypes.md [hunk 2]
```diff
@@ -0,0 +1,757 @@
...
+  a type-level natural number to bound the scopes of bound variables, as
+  shown in [this
+  implementation](https://github.com/sweirich/lennart-lambda/blob/master/lib/DeBruijnScoped.lhs).
+
+- [Finger trees](http://www.staff.city.ac.uk/~ross/papers/FingerTree.html)
+
+  Haskell's implementation of the
+  [sequence](https://hackage.haskell.org/package/containers-0.6.4.1/docs/Data-Sequence.html)
+  data structure is built on FingerTrees. In the module
+  [DFinger.lhs](./assets/DFinger.lhs) I've sketched out a nat-indexed replacement
```
Comment (sweirich) [inline blog/2020-12-04-nested-datatypes/2020-12-04-nested-datatypes.md:743]: Better to link to the dth repo itself. That file has a link to other files in the same repo (along with a cabal file etc).

**Review by sweirich — CHANGES_REQUESTED** [review]

File: blog/2020-12-04-nested-datatypes/2020-12-04-nested-datatypes.md [hunk 3]
```diff
@@ -0,0 +1,757 @@
+---
+title: Do we need nested datatypes?
+author: Stephanie Weirich
+tags: haskell
+description: An exploration of nested types and perfect trees in Haskell.
+---
+
+The following was generated from a [literate Haskell file](./assets/Nested.lhs).
```
Comment (Lysxia) [inline blog/2020-12-04-nested-datatypes/2020-12-04-nested-datatypes.md]:
```suggestion
This blog post was generated from a [literate Haskell file](./assets/Nested.lhs).
```

File: blog/2020-12-04-nested-datatypes/2020-12-04-nested-datatypes.md [hunk 4]
```diff
@@ -0,0 +1,757 @@
...
+
+[^0]: Many examples of nested datatypes, especially for perfect trees, use the type
+  `(a,a)` instead of `Two`. However, it is convenient in modern Haskell to
+   have the appropriate definitions of `fmap` etc. available for this
+  auxiliary type.
+[^1]: Regular datatypes produce regular trees, i.e. those that are either finite or self similar. Since the collection of perfect trees is a subset of normal binary trees (which are regular) it is a bit confusing that we use a non-regular datatype to represent them. But, note that this nested datatype definition includes a height "prefix" that prevents it from being self-similar (i.e. we cannot embed smaller values of this type directly into larger values). 
+[^2]: Fritz Henglein, Type Inference with Polymorphic Recursion. ACM Transactions
+  on Programming Languages and Systems. Vol 15, Issue 2. April 1993.
+[^3]: Assaf J Kfoury, Jerzy  Tiuryn, Paweł Urzyczyn. Type reconstruction in the presence of polymorphic recursion. ACM Transactions on Programming Languages and Systems. Vol 15, Issue 2. April 1993.
+[^4]: https://www.haskell.org/onlinereport/decls.html#type-signatures
```
Comment (Lysxia) [inline blog/2020-12-04-nested-datatypes/2020-12-04-nested-datatypes.md]: This and the URL in footnote 6 aren't clickable.

File: blog/2020-12-04-nested-datatypes/2020-12-04-nested-datatypes.md (hunk 2 above)
Comment (hgoldstein95) [reply blog/2020-12-04-nested-datatypes/2020-12-04-nested-datatypes.md]: Done.

**Review by Lysxia — COMMENTED** [review]
I did a pass to check formatting. No issues there. I left two comments.

**Comment by hgoldstein95** [comment]
I think we're good, but let's hold off until Monday morning to post. I don't think anyone is going to want something new to do at 3pm on a Friday.

**Comment by hgoldstein95** [comment]
@dunnl How do we feel about merging this?

**Comment by dunnl** [comment]
@hgoldstein95 We feel great about it! I gave it a quick scan and don't have anything to add. I'll merge and if you want to deploy it just give me the word.

**Pull request merged**